### ISSUE TITLE
How to get the loss of `lagrangian_regularization`

### ISSUE THREAD

**Issue by CaffreyR** [princeton-nlp/CoFiPruning#31]
Hi! In your code you calculate the `Lc`
<img width="439" alt="image" src="https://user-images.githubusercontent.com/84232793/196970087-518d8dd9-e948-49c8-85ca-9749b2375675.png">

https://github.com/princeton-nlp/CoFiPruning/blob/main/trainer/trainer.py#L682

And you use `expected_size` to calculate `expected_sparsity` , but does it match the equation in your paper?
<img width="414" alt="image" src="https://user-images.githubusercontent.com/84232793/196970790-dbc26865-bddc-4900-9f75-56e35f6622db.png">

https://github.com/princeton-nlp/CoFiPruning/blob/main/models/l0_module.py#L267

Actually you said that `sˆ is the expected model sparsity calculated from z` , but the lagrangian_regularization() do not have `inputs` or `z`
Many thanks!

**Comment by zhangzhenyu13** [comment]
z is achieved via teh hard-sigmoid func, via (1- Q(z| tehta)), i.e. 1- cdf_qz(0), which is a sampled score in forward.

Therefore, hat(s) is not achieved by inputs but only by the parameters such as loga, etc. (updated during training) .

**Comment by CaffreyR** [comment]
Emm, does it relevent to the equation of the paper? I mean in the paper it did say `sˆ ` in these two equation
![image](https://user-images.githubusercontent.com/84232793/197792457-17b15af4-46b9-4b7f-9cd2-937e856ea133.png)
<img width="474" alt="image" src="https://user-images.githubusercontent.com/84232793/197792554-7e7de063-bede-47c2-859e-bca717fd93ba.png">

**Comment by CaffreyR** [comment]
And what is the difference of `cdf_qz` and `quantile_concrete` function?

**Comment by zhangzhenyu13** [comment]
check the paper carefully, the L0 norm is proposed here: [Learning Sparse Neural Networks through $L_0$ Regularization (arxiv.org)](https://arxiv.org/abs/1712.01312)
detailed description of q(z) and that Q(z) the CDF of q is derived in the paper, where z is generated exactly as the paper in CoFi and the difference is that CoFi applies structured grouped parameter-masking strategies.

**Comment by xiamengzhou** [comment]
Thanks @zhangzhenyu13, for the answer! 

It's also worth checking out [Structured Pruning of Large Language Models](https://arxiv.org/pdf/1910.04732.pdf), which is the first work that proposes to adapt L0 regularization to control the sparsity of the models.
